docs/RFCS/20220628_invisible_index.md

```diff
@@ -0,0 +1,212 @@
+- Feature Name: Invisible Index
+- Status: draft
+- Start Date: 2022-06-28
+- Authors: Wenyi Hu
+- RFC PR: (TODO (wenyihu6): link PR later)
+- Cockroach Issue: https://github.com/cockroachdb/cockroach/issues/72576,
+  https://github.com/cockroachdb/cockroach/issues/82363
+
+# Summary
+
+This new feature introduces the option to make an index become invisible. An
+invisible index is an index that is up-to-date but is ignored by the optimizer
+unless explicitly specified with [index
+hinting](https://www.cockroachlabs.com/docs/v22.1/table-expressions#force-index-selection).
+Users can create an index as invisible or alter an index to be invisible after
+its initialization. As for now, primary indexes cannot be invisible. But unique 
+indexes can still be invisible. Specifically, the unique constraint still prevents 
+insertion of duplicates into a column regardless of whether the inedx is invisible.
+But the index will be ignored by the optimizer for queries.
+
+The main purpose of this RFC is to introduce the feature, to document different 
+choices of potential SQL syntaxes, and ultimately to justify the decision.
+
+# Motivation
+
+Currently, users are not able to observe the impact of removing an index without
+risking the cost of rebuilding the index. This new feature would allow users to
+validate whether an index should be dropped by changing it to invisible first.
+If a drop in query performance is observed, the index can be quickly toggled
+back to visible without rebuilding the index.
+
+Similarly, this new feature would also allow users to roll out new indexes with
```

Comment:
Hmm, not sure whether this fully solves this portion of the motivation. You still need to do a backfill to introduce the index, and then pay the price on every write. I do buy that it does let you experiment with seeing if READ latencies improve with the index though (just wonder if it's worth the write cost!). Am I missing something?

Reply:
I believe you are right. We are just marking the index as invisible for queries like SELECT. The indexes are maintained up to date with every writes.

Reply:
Any time the set of indexes changes, query plans can change too. It's easy to imagine cases where dropping an index negatively affects query plans. It's harder to imagine cases where adding an index leads to worse query plans, but it's possible. This paragraph can be more clear that this RFC is solving for the impact of adding an index **on query plans**, not the impact on index backfills and maintenance during mutations.